Type constraint index tuning (#712)
neo4j/docs-cypher: modules/ROOT/pages/query-tuning/indexes.adoc

|===

[NOTE]
====

> The above set of predicate can be extended with the use of type constraints, see xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[Extending index compatibility with type constraints].

Reviewer: ```suggestion
The above set of predicates can be extended with the use of type constraints.
For more information, see xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[Extending index compatibility with type constraints].
```

|===

[NOTE]
====

> The above set of predicates can be extended with the use of type constraints.

Reviewer: ```suggestion
As of Neo4j 5.12, The above set of predicates can be extended with the use of type constraints.
```

Author: ```suggestion
As of Neo4j 5.11, the above set of predicates can be extended with the use of type constraints.
```

|===

[NOTE]
====

> The above set of predicate can be extended with the use of type constraints, see xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[Extending index compatibility with type constraints].

Reviewer: ```suggestion
The above set of predicates can be extended with the use of type constraints, see xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[Extending index compatibility with type constraints].
```

|===

[NOTE]
====

> The above set of predicates can be extended with the use of type constraints, see xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[Extending index compatibility with type constraints].

Reviewer: ```suggestion
As of Neo4j 5.12, the above set of predicates can be extended with the use of type constraints, see xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[Extending index compatibility with type constraints].
```

> The above set of predicates can be extended with the use of type constraints, see xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[Extending index compatibility with type constraints].

Author: ```suggestion
As of Neo4j 5.11, the above set of predicates can be extended with the use of type constraints, see xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[Extending index compatibility with type constraints].
```

`MATCH (n:Label) WHERE $param STARTS WITH '' AND n.prop = $param`
* If expressions which are not of type string should be converted to string, then wrapping these in `toString(<expression>)` is the right choice:
`MATCH (n:Label) WHERE n.prop = toString($param)`

> * If it's known that the property is always of type string, then it's also possible to xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[add a type constraint to help the planner].

Reviewer: ```suggestion
* If it is known that the property is always of type `STRING`, then it's also possible to xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[add a type constraint to help the planner].
```

[[extending-index-compatibility-with-type-constraints]]
== Extending index compatibility with type constraints

> As explained in above sections, for indexes that are compatible only with specific types (i.e. `TEXT` and `POINT` indexes) the planner needs to be able to deduce that a predicate will evaluate to `null` for non-compatible values.

Reviewer: ```suggestion
For indexes that are compatible only with specific types (i.e. `TEXT` and `POINT` indexes), the planner needs to be able to deduce that a predicate will evaluate to `null` for non-compatible values.
```

Since xref::constraints/index.adoc#types-of-constraint[type constraints] guarantee that a property is always of the same type, they can be used to extend the scenarios for which `TEXT` indexes and `POINT` indexes are compatible with a predicate.

> For example if the property `prop` in below clause has been constrained to have type `String`, then a `TEXT` index can also be planned for the `IS NOT NULL` predicate:

Reviewer: ```suggestion
For example, if the property `prop` in the below query has been constrained to have type `STRING`, then a `TEXT` index can also be planned for the `IS NOT NULL` predicate:
```

[source]
----
MATCH (n: Label) WHERE n.prop IS NOT NULL
----

> Similarly, if the property had been constrained to have type `Point`, then a `POINT` index could have been used.

Reviewer: ```suggestion
Similarly, if the property had been constrained to have the type `POINT`, then a `POINT` index could have been used.
```

`MATCH (n:Label) WHERE $param STARTS WITH '' AND n.prop = $param`
* If expressions which are not of type string should be converted to string, then wrapping these in `toString(<expression>)` is the right choice:
`MATCH (n:Label) WHERE n.prop = toString($param)`
* If it is known that the property is always of type `STRING`, then it's also possible to xref::query-tuning/indexes.adoc#extending-index-compatibility-with-type-constraints[add a type constraint to help the planner].

[[extending-index-compatibility-with-type-constraints]]

> == Extending index compatibility with type constraints

Author: ```suggestion
== Extending index compatibility with type constraints

_This feature was introduced in Neo4j 5.11._

```

> == Extending index compatibility with type constraints
> 

Reviewer: ```suggestion
== Extending index compatibility with type constraints

_This feature was introduced in Neo4j 5.12._
```
